### ISSUE TITLE
dataclasses requiring InitVar/ClassVar at runtime

### ISSUE THREAD

**Issue by jonyscathe** [snok/flake8-type-checking#83]
As per https://docs.python.org/3/library/dataclasses.html#class-variables and https://docs.python.org/3/library/dataclasses.html#init-only-variables dataclass() inspects the type of a field in order to treat those types in some special way.

Code like their example (with imports added):

```python
from dataclasses import InitVar, dataclass

@dataclass
class C:
    i: int
    j: int = None
    database: InitVar[DatabaseType] = None

    def __post_init__(self, database):
        if self.j is None and database is not None:
            self.j = database.lookup('j')

c = C(10, database=my_database)
```

Will throw TC002 on the import of InitVar.
But if InitVar is moved into a type checking block then initializing C will fail because database won't exist in the __post_init__ function.

I imagine the only way to handle this (other than my current solution of a noqa anywhere I use InitVar) will to add exceptions for InitVar and ClassVar if dataclass is also imported or something along those lines ?

**Comment by sondrelg** [comment]
I agree, we should add handling for this 👍 

Do you think it makes sense to just ignore `InitVar`, or are there cases where dataclasses will evaluate other annotations? 

Would you be interested in creating a fix for this?

**Comment by jonyscathe** [comment]
From the dataclasses documentation it appears that it will evaluate `InitVar` or `ClassVar` annotations.
I can try making a fix for this.

**Comment by sondrelg** [comment]
I started looking at this, but I'm no longer able to replicate the issue after v2. 

v2 changes the plugin behaviour so that we won't flag imports like `from dataclasses import InitVar` if there's already one or more other imports from the module made that cannot be guarded. This means, since `from dataclasses import dataclass` will always exist in the same file, the `InitVar` import should never be flagged. The only way to mess up is if you did something like this:

```python
from my_app.some_file import dataclass  # bad import
from dataclasses import InitVar
```

But this doesn't seem worth handling.

The `typing` module was also completely whitelisted in v2 since we assume `from typing import TYPE_CHECKING` will be needed in most scopes.

Are you able to still replicate the issue? If not, I think we might consider this resolved, accidentally 🙂

**Comment by sondrelg** [comment]
I'll close this for now. Please let me know if you still think there's an issue 👍

**Comment by jonyscathe** [comment]
Sorry I was sick yesterday and wasn't at my computer.
I checked this morning, and yes, it looks like this has been fixed :+1:

**Comment by sondrelg** [comment]
No worries! Thanks for checking 🙂
